### PR TITLE
test(helpers): consolidate test-fixture scaffolding into test/_helpers.mjs (closes #1700)

### LINKED ISSUE #1700: Test fixture scaffolding consolidation in test/_helpers.mjs
## Context

Child 11 of the simplification epic (#1689). Source: whole-repo simplification audit, adversarially verified by a 13-agent workflow on 2026-08-12 (main at d396f33b), re-verified against main 81e86581 on 2026-08-17. This issue removes or consolidates test scaffolding without losing coverage.

## Scope (verifier-confirmed)

- Promote withTempDir(async (dir) => {}) into test/_helpers.mjs. The prototype lives at test/loop/inspect-run-test-helpers.mjs:99-106 (mkdtemp + try/finally rm, hardcoded prefix "pi-inspect-run-test-"). The promoted helper takes an optional prefix; the prototype file re-exports from ../_helpers.mjs or its callers import _helpers directly, so one definition remains.
- Adopt withTempDir incrementally at the ~1195 hand-rolled mkdtemp/mkdtempSync call sites (138 test files), starting with the two heaviest files: test/github/upsert-checkpoint-verdict.test.mjs (95 sites; note it lives under test/github, not test/loop) and test/loop/resolve-dev-loop-startup.test.mjs (66 sites).
- Add captureStream() to test/_helpers.mjs and delete the 10 byte-identical local copies in test/github (comment-issue, create-issue, edit-comment, edit-issue, edit-pr, fetch-ci-logs, list-issues, tick-verified-checkboxes, view-issue, view-pr).
- Add one shared entries-based gh stub with a recorded calls array to test/_helpers.mjs. The 10 local stubGh definitions in the same files carry 4 distinct signatures (sequence-of-responses: comment-issue:15, edit-comment:15, tick-verified-checkboxes:12; fixed-response: create-issue:10, edit-issue:11, edit-pr:11; payload+opts: view-issue:6, view-pr:6, list-issues:10; matchers: fetch-ci-logs:12). Files with divergent signatures keep one-line local wrappers over the shared stub. The domain-routing stubs in write-gate-context.test.mjs (lines 59, 225) stay local. First evaluate whether the existing makeGhMock (test/_helpers.mjs:49) already covers these cases before adding a new export.
- Add initGitFixture(dir, { remote, branch, commit }) covering the common git init/identity/commit/optional-remote block. ~24 test files under test/loop and test/github contain such blocks; convert only blocks that match the common shape exactly. Exotic fixtures stay local: ensure-worktree.test.mjs builds a self-referential origin plus "remote set-head origin --auto" for git < 2.49 compat (~lines 49-63) and clone-based two-repo shapes (~lines 66-88).

## Required accompanying edits

- None beyond scope items

## Tests and validators that must stay green

- npm run verify (test:scripts covers test/github/*.test.mjs and test/loop/*.test.mjs; converted files must pass unchanged)
- test/contracts/harness-pi-import-boundary, cli-harness-agnostic, skills-no-cross-package-relative-import, no-package-escaping-imports (scanner unification is a non-goal; these must pass untouched)

## Dependencies

- None

## Risk notes

Sweep-only and wide blast radius: do NOT convert all ~1195 sites in one PR. Ship the helpers plus conversion of the 2 heaviest files as the epic child; further adoption is opportunistic per-file cleanup. Keep writeGhStub (PATH/file-based, test/_helpers.mjs:300) distinct from the new in-memory stub, and check makeGhMock for overlap before adding a new export. The import-boundary contract tests each embed their own scanner with test-specific edge-case coverage (dynamic and side-effect import detection, harness-pi-import-boundary.test.mjs:75-86); leave them alone.

Estimated lines removed: ~1500 (directional).

## Acceptance criteria

- [x] test/_helpers.mjs exports withTempDir(fn, { prefix }) that mkdtemps under os.tmpdir(), awaits fn(dir), and always rm -rf's in finally; behavior matches the prototype at test/loop/inspect-run-test-helpers.mjs:99-106
- [x] test/loop/inspect-run-test-helpers.mjs no longer defines its own withTempDir (re-exports from ../_helpers.mjs or its callers import _helpers directly); the inspect-run suites pass unchanged
- [x] test/_helpers.mjs exports captureStream(); all 10 identical local copies in test/github/*.test.mjs are deleted and replaced with the import
- [x] test/_helpers.mjs exports one shared entries-based gh stub with a recorded calls array (a new export, or a documented extension of the existing makeGhMock at test/_helpers.mjs:49 if it already fits); the local stubGh definitions in comment-issue, create-issue, edit-comment, edit-issue, edit-pr, fetch-ci-logs, list-issues, tick-verified-checkboxes, view-issue, and view-pr tests route through it (thin per-file wrappers allowed where signatures differ); write-gate-context's routing stubs (lines 59, 225) may stay local
- [x] writeGhStub (test/_helpers.mjs:300) and makeGhMock remain unchanged in behavior; the new in-memory stub is documented as distinct from writeGhStub
- [x] test/_helpers.mjs exports initGitFixture(dir, { remote, branch, commit }) covering the common init/identity/commit/remote block; only blocks matching that shape are converted, and exotic fixtures (self-referential origin + set-head, clone-based two-repo setups in ensure-worktree.test.mjs) stay local
- [x] test/github/upsert-checkpoint-verdict.test.mjs and test/loop/resolve-dev-loop-startup.test.mjs are converted to withTempDir; no repo-wide mass conversion of the remaining ~1100 mkdtemp sites in this PR
- [x] No test semantics change: converted files keep the same assertions and exit-code expectations, and npm run verify passes
- [x] At pickup, re-verify file paths, line pins, and site counts against current main (sibling #1692 may still move package.json scripts); shrink scope where the ground moved

## Definition of done

- [x] npm run verify passes (test:scripts at package.json:29 covers test/github/*.test.mjs and test/loop/*.test.mjs; no test:loop or test:github npm scripts exist)
- [x] node --test test/contracts/harness-pi-import-boundary.test.mjs test/contracts/cli-harness-agnostic.test.mjs test/contracts/no-package-escaping-imports.test.mjs test/contracts/skills-no-cross-package-relative-import.test.mjs passes with those files untouched
- [x] grep confirms zero remaining local captureStream definitions under test/github/
- [x] grep confirms zero remaining local stubGh definitions in the 10 converted test/github files (write-gate-context routing stubs exempt)
- [x] Net diff in converted files is negative (~1500 lines removed is directional, not a hard gate)
- [x] PR body records the re-verification result against current main at pickup

## AC / DoD matrix

| Acceptance criterion | Verified by DoD item(s) |
|---|---|
| withTempDir exported from test/_helpers.mjs, matches prototype behavior | npm run verify (inspect-run suites pass unchanged) |
| inspect-run-test-helpers.mjs re-exports withTempDir instead of defining it | npm run verify |
| captureStream added, 10 local copies deleted | grep confirms zero remaining captureStream defs |
| shared entries-based gh stub added, 10 stubGh defs route through it | grep confirms zero remaining stubGh defs (write-gate-context exempt) |
| writeGhStub/makeGhMock unchanged, new stub documented distinct | npm run verify |
| initGitFixture added, only matching blocks converted, exotic fixtures stay local | npm run verify |
| upsert-checkpoint-verdict.test.mjs & resolve-dev-loop-startup.test.mjs converted, no mass conversion | npm run verify; contract tests (import-boundary etc.) untouched |
| no test semantics change, npm run verify passes | npm run verify; net diff negative in converted files |
| file paths/line pins/site counts re-verified at pickup | PR body records re-verification result |

## Non-goals

- Converting all ~1195 mkdtemp sites repo-wide (helpers + the 2 heaviest files only; further adoption is opportunistic per-file cleanup)
- makeSuiteHelpers(scriptPath) factory for the 4 *-test-helpers.mjs files (saves ~5 lines per file; unique fixture content dominates)
- makeCheckpointFixture(overrides) builder (optional in the original audit; deferred)
- Unifying the 4 import-boundary contract-test scanners (each embeds test-specific edge-case coverage such as dynamic/side-effect import detection; risk of coverage loss)
- Source-side gh helper creation or migration (owned by #1695, #1696, #1697)
- scripts/github gate-tooling dedup (owned by #1698)
- packages/core/src/loop production helper dedup (owned by #1699)
- Any change to production code under scripts/ or packages/

## Scope boundary

This issue owns shared test fixture scaffolding in test/_helpers.mjs (withTempDir, captureStream, an entries-based gh stub, initGitFixture) plus conversion of the 2 heaviest test files. It does NOT own production-side helper dedup: the shared gh helper (#1695, #1697), scripts/github gate tooling (#1698), or packages/core/src/loop helpers (#1699).

## Re-verification (2026-08-17)

Verified against main 81e86581. Drift since d396f33b:

- mkdtemp/mkdtempSync site count grew from ~1130 call lines in 132 files to 1195 in 138 files (body originally said ~979). Heavy-file counts: upsert-checkpoint-verdict.test.mjs 94 to 95, resolve-dev-loop-startup.test.mjs 65 to 66. Both remain the two heaviest; first-conversion targets unchanged.
- test/_helpers.mjs grew to 353 lines: makeGhMock moved from line 16 to line 49, writeGhStub from line 267 to line 300. makeGhMock now has 7 test-file importers.
- write-gate-context.test.mjs routing stubs moved from lines 58/224 to 59/225.
- The 10 captureStream copies and 10 stubGh copies in test/github persist with the same 4 signature groups; some captureStream pins shifted by a few lines (e.g. comment-issue 23 to 26). No new copies appeared: the comment-body-guard work on comment-issue/edit-comment/create-pr surfaces did not add stubs, and the checkpoint-evidence rework did not displace upsert-checkpoint-verdict.test.mjs.
- git-init fixture files count 25 to 24. ensure-worktree.test.mjs exotic fixtures now span roughly lines 49-63 (self-referential origin + set-head) and 66-88 (clone-based).
- harness-pi-import-boundary.test.mjs dynamic/side-effect scanner assertions now sit at lines 75-86 (was cited as 75-80).
- package.json pins unchanged: verify at line 24, test:scripts at line 29. The rc.6 milestone shipped; the stale rc.6 issue list was dropped from the pickup re-check criterion. All scope items remain undone on main.

### PR DESCRIPTION
**Pull request by mfittko**
## Objective

Child 11 (final) of the simplification epic #1689: consolidate duplicated test-fixture scaffolding into `test/_helpers.mjs` and delete the copies. Test-only, no production change. Net **-534 lines** (22 files, +424 / -958; the +6 vs the original consolidation are the repo-local git-identity lines added by the round-2 `initGitFixture` fix, net of the condensed inline comment from the JSDoc fix).

Closes #1700.

## Re-verification at pickup (against main 2f886fc0)

Paths/counts re-checked. `test/_helpers.mjs` `makeGhMock`/`writeGhStub` present as pinned; the 10 `captureStream` + 10 `stubGh` copies in `test/github/` still present; the two heaviest mkdtemp files (`upsert-checkpoint-verdict.test.mjs`, `resolve-dev-loop-startup.test.mjs`) remain the heaviest. All scope items were undone on main.

## In scope (done)

- **`withTempDir(fn, { prefix })`** promoted into `test/_helpers.mjs` (mkdtemp under `os.tmpdir()`, `await fn(dir)`, always `rm -rf` in `finally`). `test/loop/inspect-run-test-helpers.mjs` now re-exports it bound to the prototype's `"pi-inspect-run-test-"` prefix — one definition remains.
- **`captureStream()`** added; the 10 byte-identical copies in `test/github/` (comment-issue, create-issue, edit-comment, edit-issue, edit-pr, fetch-ci-logs, list-issues, tick-verified-checkboxes, view-issue, view-pr) deleted and imported.
- **`makeGhStub`** (new, entries-based, with a recorded `calls` array; order-based + predicate-routed). `makeGhMock` was evaluated first and does NOT fit — its `runChild`/4-arg signature, encoded-exit-code overflow (97) vs the local stubs' throw-on-overflow, and lack of predicate routing all differ — so a distinct stub was added and documented as separate from both `makeGhMock` and `writeGhStub` (both unchanged). The 10 `test/github` `stubGh` defs now route through it via thin per-file wrappers where signatures differ. `write-gate-context.test.mjs`'s routing stubs stay local.
- **`initGitFixture(dir, { remote, branch, commit })`** added (common init/identity/commit/optional-remote). Converted 8 files matching the exact common shape (`test@example.com`/`Test` identity, plain init+commit(+remote)). Exotic fixtures left local by design: different identity literals, custom `GIT_CONFIG_GLOBAL/SYSTEM` isolation, `gpgsign false`, `symbolic-ref`-based, and `ensure-worktree.test.mjs`'s self-referential-origin + clone-based two-repo shapes.
- **Heavy-file conversion**: `upsert-checkpoint-verdict.test.mjs` (101/105 sites) and `resolve-dev-loop-startup.test.mjs` (46/55) converted to `withTempDir`; the remaining ~9 are exotic multi-resource/dual-repo shapes left local. No mass conversion of the other ~1100 sites (opportunistic per-file cleanup later).

## Line-count note

Net -534, short of the ~1500 directional estimate. The shortfall is deliberate scope discipline: `initGitFixture` converted only exact-shape blocks (many git fixtures use non-default identities or custom envs and were correctly left local rather than force-converted), and the ~9 exotic mkdtemp sites in the two heavy files stayed local. Forcing those would change fixture semantics. The estimate was explicitly labeled "directional, not a hard gate."

## Verification

- `npm run verify` — full green: assets 285, extension 126, pack 1, dev-loop 38, core 2798, scripts 4167 pass / 36 skipped / 0 fail, docs + workflows clean. Ran twice, identical.
- The 4 import-boundary contract tests pass standalone (27/27), untouched.
- grep: zero remaining local `captureStream` defs under `test/github/`; the 10 `stubGh` defs now route through `makeGhStub` (write-gate-context routing stubs exempt).
- No production code touched (`git diff --name-only` is entirely under `test/`); no test-semantics change (identical assertions/exit-code expectations throughout).

## Acceptance criteria

- [x] `withTempDir(fn, { prefix })` exported from `test/_helpers.mjs`, behavior matches the prototype.
- [x] `inspect-run-test-helpers.mjs` no longer defines its own `withTempDir` (re-exports); inspect-run suites pass unchanged.
- [x] `captureStream()` exported; all 10 local copies deleted and replaced with the import.
- [x] One shared entries-based gh stub with a recorded `calls` array (new `makeGhStub`, documented distinct after evaluating `makeGhMock`); the 10 `stubGh` defs route through it (thin wrappers where signatures differ); write-gate-context routing stubs stay local.
- [x] `writeGhStub` and `makeGhMock` unchanged; the new in-memory stub documented as distinct.
- [x] `initGitFixture` exported; only exact-shape blocks converted; exotic fixtures stay local.
- [x] The two heaviest files converted to `withTempDir`; no repo-wide mass conversion.
- [x] No test semantics change; `npm run verify` passes.
- [x] Pickup re-verification recorded above.

## Definition of done

- [x] `npm run verify` passes.
- [x] The 4 contract tests pass with those files untouched.
- [x] grep confirms zero remaining local `captureStream` defs under `test/github/`.
- [x] grep confirms zero remaining local `stubGh` defs in the 10 converted files (write-gate-context exempt).
- [x] Net diff in converted files is negative (net -534; ~1500 was directional, not a hard gate — see note).
- [x] PR body records the pickup re-verification.

## Non-goals

- Converting all ~1195 mkdtemp sites repo-wide (helpers + 2 heaviest files only).
- `makeSuiteHelpers` factory / `makeCheckpointFixture` builder (deferred per the issue).
- Unifying the 4 import-boundary scanners (coverage-loss risk).
- Any production-side dedup (#1695/#1696/#1697/#1698/#1699) or production code change.

## AC / DoD matrix

| Acceptance criterion | Verified by |
|---|---|
| withTempDir exported, matches prototype | `npm run verify` (inspect-run suites unchanged) |
| inspect-run-test-helpers re-exports withTempDir | `npm run verify` |
| captureStream added, 10 copies deleted | grep zero remaining defs |
| shared gh stub added, 10 stubGh route through it | grep zero remaining stubGh defs (write-gate-context exempt) |
| writeGhStub/makeGhMock unchanged, new stub distinct | `_helpers.mjs` diff additive; `npm run verify` |
| initGitFixture added, only matching blocks converted | `npm run verify` |
| 2 heaviest files converted, no mass conversion | `npm run verify`; contract tests untouched |
| no test semantics change | `npm run verify`; net -534 |
| pickup re-verification recorded | this PR body |

### CODE REVIEW

**Review by mfittko — COMMENTED** [review]
### Gate review: `draft_gate`


**Reviewed head SHA:** `d5597c28149234a0402e2c4a6e30447078ddc8c8`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 5 angles reviewed; 1 finding (see per-angle breakdown below).

- `correctness` → `clean`
- `coverage` → `clean`
- `determinism` → `clean`
- `gate-evidence` → `clean`
- `pr-description` → `findings_present` (1 finding)

**Body-filed findings** (no in-diff location):


> **nit** (`pr-description`): Line-count figures are stale after the round-2 identity fix: body states '+418 / -958, net -540' but the authoritative diff is +426 / -958, net -532. Deletions (-958) and file count (22) still match; the +8 insertions are the repo-local git-identity lines added this round.
> Location: `PR body`

**Next action:** Draft gate round 2 clean at d5597c28 (4 dispatch units): the round-1 HIGH initGitFixture hermetic-identity regression is fixed (repo-local user.email/user.name persisted) and re-verified hermetically by 3 reviewers; CI green; 1 nit (PR-body line count, since corrected). Proceeding to ready-for-review.

File: test/_helpers.mjs [hunk 1]
```diff
@@ -182,6 +240,30 @@ function runGitFixture(cwd, args) {
   return result.stdout.trim();
 }
 
+/**
+ * Common `git init` + identity + optional initial commit + optional remote
+ * fixture. Identity comes from GIT_FIXTURE_ENV's author/committer overrides
+ * (the same "Test <test@example.com>" every explicit `git config user.*` in
+ * these suites already resolves to), so no separate config calls are needed.
+ * @param {string} dir — an existing empty directory (e.g. a mkdtemp result)
```
[the remark this answers is not in the document]
Comment (mfittko) [reply test/_helpers.mjs:250]: Fixed: corrected the initGitFixture JSDoc to state that identity is written into repo-local config after init (so it persists for env-less callers on a hermetic runner), removing the stale claim that GIT_FIXTURE_ENV alone suffices with no config calls. Condensed the inline comment to reference the JSDoc. Doc-only change.

**Review by mfittko — COMMENTED** [review]
### Gate review: `pre_approval_gate`


**Reviewed head SHA:** `08272b10be4d339caa96e1695d0c3080855e75b3`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 5 angles reviewed; 1 finding (see per-angle breakdown below).

- `acceptance-criteria` → `clean`
- `contradiction-lens` → `clean`
- `pr-checklist-matrix` → `findings_present` (1 finding)
- `renderer-security` → `clean`
- `yagni` → `clean`

**Body-filed findings** (no in-diff location):


> **nit** (`pr-checklist-matrix`): Matrix line-count figures are off by 2: PR body states +426 / -958 (net -532) and derives a "Line-count note" plus a "+8 vs the original consolidation" claim, but the authoritative counts (gh PR API and git diff --stat vs merge-base, in agreement) are +424 / -958 (net -534); the +8 delta is actually +6. Net-negative DoD still holds, so this is cosmetic accuracy only.
> Location: `PR body`

**Gate evidence note:** Copilot review rounds exhausted (2/2); current head has zero unresolved threads and green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Pre-approval fan-out clean at 08272b10 (3 dispatch units, 5 angles): all clean bar 1 nit (PR-body line count, since corrected to +424/-958 net -534). JSDoc now matches the identity-persistence fix; Copilot round-1 finding resolved, internal-tooling PR (Copilot suppressed on current head); CI green; ACs/DoD ticked. Proceeding to merge.
